src/FuncSharp/Option/Option.cs

```diff
@@ -220,12 +220,34 @@ public override int GetHashCode()
             return Structural.HashCode(NonEmpty, Value);
         }
 
+        public static bool operator ==(Option<A> obj1, object obj2)
+        {
+            return ReferenceEquals(obj1, null)
+                ? ReferenceEquals(obj2, null)
+                : obj1.Equals(obj2);
+        }
+
+        public static bool operator !=(Option<A> obj1, object obj2)
+        {
+            return ReferenceEquals(obj1, null)
+                ? !ReferenceEquals(obj2, null)
+                : !obj1.Equals(obj2);
+        }
+
         public override bool Equals(object obj)
         {
             if (obj is IOption<A> other)
             {
                 return NonEmpty == other.NonEmpty && Equals(Value, other.GetOrDefault());
             }
+            if (typeof(A) == typeof(NonEmptyString) && obj is IOption<string> otherString)
```

Comment:
how about moving these to separate `Equals` methods and implement `IEquatable<NonEmptyString>` for instance?

Reply:
I don't think that every `Option<A>` should be comparable to `Option<NonEmptyString>`

i can move it to another meththod though.

Reply:
Yeah, I  tried, but breaking it into a separate method makes sit weird, because we're still inside an `Option<A>`, not `Option<string>` and casting `this` seems weird to me. It's not that many lines and not so hard to read. IMO

Reply:
Best I could do is this: 

```
public override bool Equals(object obj)
{
    if (obj is IOption<A> other)
    {
        return NonEmpty == other.NonEmpty && Value.SafeEquals(other.GetOrDefault());
    }
    if (this is Option<NonEmptyString> nonEmptyStringOption && obj is IOption<string> otherString)
    {
        return StringEqualsWithNonEmptyString(otherString, nonEmptyStringOption);
    }
    if (this is Option<string> stringOption && obj is IOption<NonEmptyString> otherNonEmptyString)
    {
        return StringEqualsWithNonEmptyString(stringOption, otherNonEmptyString);
    }
    return false;
}

public static bool StringEqualsWithNonEmptyString(IOption<string> first, IOption<NonEmptyString> other)
{
    return first.NonEmpty == other.NonEmpty && string.Equals(other.GetOrDefault(), first.GetOrDefault());
}
```

Reply:
It is not about number of lines, but being explicit in terms of "what do you actually copmare the IOption with". So what I meant was to implement `IEquatable<IOption<string>>` for example.

Reply:
Right, but I don't want an Option<int> to have an `Equals(Option<NonEmptyString> other)` method. I don't think it should be there.